plugins/file_selector/file_selector_linux/linux/file_selector_plugin.cc

```diff
@@ -0,0 +1,238 @@
+// Copyright 2020 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "include/file_selector_linux/file_selector_plugin.h"
+
+#include <flutter_linux/flutter_linux.h>
+#include <gtk/gtk.h>
+
+// From method_channel_file_selector.dart
+const char kChannelName[] = "plugins.flutter.io/file_selector";
+
+const char kOpenFileMethod[] = "openFile";
+const char kGetSavePathMethod[] = "getSavePath";
+const char kGetDirectoryPathMethod[] = "getDirectoryPath";
+
+const char kAcceptedTypeGroupsKey[] = "acceptedTypeGroups";
+const char kConfirmButtonTextKey[] = "confirmButtonText";
+const char kInitialDirectoryKey[] = "initialDirectory";
+const char kMultipleKey[] = "multiple";
+const char kSuggestedNameKey[] = "suggestedName";
+
+// From x_type_group.dart
+const char kTypeGroupLabelKey[] = "label";
+const char kTypeGroupExtensionsKey[] = "extensions";
+const char kTypeGroupMimeTypesKey[] = "mimeTypes";
+
+// Errors
+const char kBadArgumentsError[] = "Bad Arguments";
+const char kNoScreenError[] = "No Screen";
+
+struct _FlFileSelectorPlugin {
+  GObject parent_instance;
+
+  FlPluginRegistrar* registrar;
+
+  // Connection to Flutter engine.
+  FlMethodChannel* channel;
+};
+
+G_DEFINE_TYPE(FlFileSelectorPlugin, fl_file_selector_plugin, G_TYPE_OBJECT)
+
+// Converts a type group received from Flutter into a GTK file filter.
+static GtkFileFilter* type_group_to_filter(FlValue* value) {
+  if (fl_value_get_type(value) != FL_VALUE_TYPE_MAP) {
+    return nullptr;
+  }
+
+  g_autoptr(GtkFileFilter) filter = gtk_file_filter_new();
+
+  FlValue* label = fl_value_lookup_string(value, kTypeGroupLabelKey);
+  if (label != nullptr && fl_value_get_type(label) == FL_VALUE_TYPE_STRING) {
+    const gchar* name = fl_value_get_string(label);
+    gtk_file_filter_set_name(filter, name);
+  }
+
+  bool has_filter = false;
+  FlValue* extensions = fl_value_lookup_string(value, kTypeGroupExtensionsKey);
+  if (extensions != nullptr &&
+      fl_value_get_type(extensions) == FL_VALUE_TYPE_LIST) {
+    for (size_t j = 0; j < fl_value_get_length(extensions); j++) {
+      FlValue* v = fl_value_get_list_value(extensions, j);
+      if (fl_value_get_type(v) != FL_VALUE_TYPE_STRING) return nullptr;
+
+      g_autofree gchar* pattern =
+          g_strdup_printf("*.%s", fl_value_get_string(v));
+      gtk_file_filter_add_pattern(filter, pattern);
+      has_filter = true;
+    }
+  }
+  FlValue* mime_types = fl_value_lookup_string(value, kTypeGroupMimeTypesKey);
+  if (mime_types != nullptr &&
+      fl_value_get_type(mime_types) == FL_VALUE_TYPE_LIST) {
+    for (size_t j = 0; j < fl_value_get_length(mime_types); j++) {
```

Comment:
Nit: should probably use `i` instead of `j` here - it's implying a nested iteration.

Reply:
Done